hydromt/models/components/grid.py

```diff
@@ -13,371 +11,163 @@
+from hydromt.models.components.model_component import ModelComponent
+from hydromt.models.root import ModelRoot
+from hydromt.workflows.basin_mask import get_basin_geometry, parse_region
+from hydromt.workflows.grid import (
+    grid_from_constant,
+    grid_from_geodataframe,
+    grid_from_raster_reclass,
+    grid_from_rasterdataset,
+    rotated_grid,
+)
+
+logger = getLogger(__name__)
+
+
+class GridComponent(ModelComponent):
+    """ModelComponent class for grid components.
+
+    This class is used for setting, creating, writing, and reading grids for a
+    HydroMT model. The grid data in the data property of this class is of the
+    hydromt.gis.raster.RasterDataset type.
+    """
 
-    def __init__(self, *args, **kwargs) -> None:
-        super().__init__(*args, **kwargs)
-        self._grid = None  # xr.Dataset()
-
-    # generic grid methods
-    def setup_grid_from_constant(
+    def __init__(
         self,
-        constant: Union[int, float],
-        name: str,
-        dtype: Optional[str] = "float32",
-        nodata: Optional[Union[int, float]] = None,
-        mask_name: Optional[str] = "mask",
-    ) -> List[str]:
-        """HYDROMT CORE METHOD: Adds a grid based on a constant value.
+        root: ModelRoot,
+        data_catalog: DataCatalog,
+        model_region=None,
+        logger: Logger = logger,
+    ):
+        """Initialize a GridComponent.
 
         Parameters
         ----------
-        constant: int, float
-            Constant value to fill grid with.
-        name: str
-            Name of grid.
-        dtype: str, optional
-            Data type of grid. By default 'float32'.
-        nodata: int, float, optional
-            Nodata value. By default infered from dtype.
-        mask_name: str, optional
-            Name of mask in self.grid to use for masking raster_fn. By default 'mask'.
-            Use None to disable masking.
-
-        Returns
-        -------
-        list
-            Names of added model grid layer.
+        root : ModelRoot
+            model root object containing path to model root
+        data_catalog : DataCatalog
+            data catalog object
+        model_region : _type_, optional
+            model region object, by default None
+        logger : Logger, optional
+            logger object, by default logger
         """
-        da = workflows.grid.grid_from_constant(
-            grid_like=self.grid,
-            constant=constant,
-            name=name,
-            dtype=dtype,
-            nodata=nodata,
-            mask_name=mask_name,
-        )
-        # Add to grid
-        self.set_grid(da)
-
-        return [name]
-
-    def setup_grid_from_rasterdataset(
+        # self._model_ref = weakref.ref(model) TODO: discuss if this is necessary
+        self.logger = logger
+        self.root = root
+        self.data_catalog = data_catalog
+        self.model_region = model_region
+        self._data = None
+
+    def set(
```

Comment:
I am wondering if this function will be called from outside the class. Perhaps it's better to make this function private and add the `create` function to the base class.

Reply:
So we should try to avoid it indeed at the ``ModelComponent`` level but at the level of ``Model`` (where my setup_reservoirs function is) I think it is fine and unavoidable that several components can get updated

Reply:
For this PR leave set as a public method. @deltamarnix if you still think this method should be private, you can open an issue and we can discuss further then.